File: frontend/src/services/memoryService.ts
import axios from 'axios';

const BASE_URL = "http://localhost:8000";
const MEMORIES_URL = `${BASE_URL}/memories`;

// Types pour la mémoire
export interface Memory {
  key: string;
  value: string;
  created_at: string;
  updated_at: string;
}

export interface UserInfo {
  name?: string;
  preferences?: {
    [key: string]: string | boolean | number;
  };
  facts?: {
    [key: string]: string;
  };
  context?: string[];
}

/**
 * Récupère toutes les mémoires associées à un utilisateur
 * @returns Liste des mémoires de l'utilisateur
 */
export const getUserMemories = async (): Promise<UserInfo> => {
  try {
    // Utiliser une promesse avec timeout pour éviter les attentes trop longues
    const fetchPromise = axios.get(`${MEMORIES_URL}/user`);
    
    // Promesse qui se résout après un délai (timeout)
    const timeoutPromise = new Promise<{data: UserInfo}>((_, reject) => {
      setTimeout(() => {
        console.warn("Timeout lors de la récupération des mémoires utilisateur");
        reject(new Error("Timeout"));
      }, 3000);
    });
    
    // Utiliser la première promesse qui se résout
    const response = await Promise.race([fetchPromise, timeoutPromise]);
    return response.data as UserInfo;
  } catch (error: any) {
    console.error("Erreur lors de la récupération des mémoires utilisateur:", error?.message || error);
    
    // Retourner un objet par défaut en cas d'erreur
    return {
      name: undefined,
      preferences: {},
      facts: {},
      context: []
    };
  }
};

/**
 * Sauvegarde une information utilisateur en mémoire
 * @param key - Clé de la mémoire (ex: "name", "preferences.theme")
 * @param value - Valeur à stocker
 * @returns Succès de l'opération
 */
export const saveUserMemory = async (key: string, value: any): Promise<boolean> => {
  // Vérifier que la clé est valide
  if (!key) {
    console.error("Clé de mémoire invalide");
    return false;
  }
  
  try {
    // Créer une promesse avec timeout pour éviter les attentes trop longues
    const savePromise = axios.post(`${MEMORIES_URL}/user`, { key, value });
    
    // Promesse de timeout pour abandonner après 3 secondes
    const timeoutPromise = new Promise<any>((_, reject) => {
      setTimeout(() => reject(new Error("Timeout lors de la sauvegarde de la mémoire")), 3000);
    });
    
    // Utiliser la première promesse qui se résout
    await Promise.race([savePromise, timeoutPromise]);
    return true;
  } catch (error: any) {
    console.error(`Erreur lors de la sauvegarde de la mémoire utilisateur (${key}):`, error?.message || error);
    return false;
  }
};

/**
 * Extrait les informations importantes d'une conversation
 * @param conversationId - ID de la conversation
 * @param messages - Messages de la conversation
 * @returns Succès de l'opération
 */
export const extractMemoriesFromConversation = async (
  conversationId: string,
  messages: { sender: string; text: string }[]
): Promise<boolean> => {
  // Vérifier que l'ID de conversation est valide
  if (!conversationId) {
    console.error("Impossible d'extraire les souvenirs: ID de conversation invalide");
    return false;
  }
  
  // Vérifier qu'il y a des messages à traiter
  if (!messages || messages.length === 0) {
    console.error("Impossible d'extraire les souvenirs: Aucun message fourni");
    return false;
  }
  
  try {
    // Créer une promesse avec timeout pour éviter les attentes trop longues
    const extractPromise = axios.post(`${MEMORIES_URL}/extract`, {
      conversation_id: conversationId,
      messages: messages.map(msg => ({
        role: msg.sender === 'user' ? 'user' : 'assistant',
        content: msg.text
      }))
    });
    
    // Promesse de timeout pour abandonner après 5 secondes
    const timeoutPromise = new Promise<any>((_, reject) => {
      setTimeout(() => reject(new Error("Timeout lors de l'extraction des souvenirs")), 5000);
    });
    
    // Utiliser la première promesse qui se résout
    await Promise.race([extractPromise, timeoutPromise]);
    return true;
  } catch (error: any) {
    console.error("Erreur lors de l'extraction des mémoires:", error?.message || error);
    return false;
  }
};

/**
 * Récupère le contexte de mémoire pour une nouvelle conversation
 * @returns Contexte de mémoire formaté pour le modèle
 */
export const getMemoryContext = async (): Promise<string> => {
  try {
    // Utiliser getUserMemories qui a déjà un timeout intégré
    const userInfo = await getUserMemories();

    
    let context = "Informations importantes sur l'utilisateur:\n";
    
    if (userInfo.name) {
      context += `- Nom: ${userInfo.name}\n`;
    }
    
    if (userInfo.preferences && Object.keys(userInfo.preferences).length > 0) {
      context += "- Préférences:\n";
      for (const [key, value] of Object.entries(userInfo.preferences)) {
        context += `  * ${key}: ${value}\n`;
      }
    }
    
    if (userInfo.facts && Object.keys(userInfo.facts).length > 0) {
      context += "- Faits importants:\n";
      for (const [key, value] of Object.entries(userInfo.facts)) {
        context += `  * ${key}: ${value}\n`;
      }
    }
    
    if (userInfo.context && userInfo.context.length > 0) {
      context += "- Contexte supplémentaire:\n";
      for (const item of userInfo.context) {
        context += `  * ${item}\n`;
      }
    }
    
    return context;
  } catch (error) {
    console.error("Erreur lors de la récupération du contexte de mémoire:", error);
    return "";
  }
};

File: Backend/app/memory_service.py
import os
import json
import datetime
from typing import Dict, List, Any, Optional, Union
import re

# Chemin vers le répertoire de stockage des mémoires
MEMORY_DIR = "./Backend/app/data/memories"

# S'assurer que le répertoire existe
os.makedirs(MEMORY_DIR, exist_ok=True)

class MemoryService:
    def __init__(self):
        self.user_memory_file = os.path.join(MEMORY_DIR, "user_memory.json")
        self._initialize_user_memory()

    def _initialize_user_memory(self):
        """Initialise le fichier de mémoire utilisateur s'il n'existe pas"""
        if not os.path.exists(self.user_memory_file):
            default_memory = {
                "name": None,
                "preferences": {},
                "facts": {},
                "context": []
            }
            with open(self.user_memory_file, 'w', encoding='utf-8') as f:
                json.dump(default_memory, f, ensure_ascii=False, indent=2)

    def _load_user_memory(self) -> Dict[str, Any]:
        """Charge la mémoire utilisateur depuis le fichier"""
        try:
            with open(self.user_memory_file, 'r', encoding='utf-8') as f:
                return json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            self._initialize_user_memory()
            return {
                "name": None,
                "preferences": {},
                "facts": {},
                "context": []
            }

    def _save_user_memory(self, memory: Dict[str, Any]):
        """Sauvegarde la mémoire utilisateur dans le fichier"""
        with open(self.user_memory_file, 'w', encoding='utf-8') as f:
            json.dump(memory, f, ensure_ascii=False, indent=2)

    def get_user_memory(self) -> Dict[str, Any]:
        """Récupère toutes les mémoires de l'utilisateur"""
        return self._load_user_memory()

    def save_user_memory(self, key: str, value: Any) -> bool:
        """
        Sauvegarde une information dans la mémoire utilisateur
        
        Args:
            key: Clé de la mémoire (ex: "name", "preferences.theme")
            value: Valeur à stocker
            
        Returns:
            bool: Succès de l'opération
        """
        memory = self._load_user_memory()
        
        # Gestion des clés imbriquées (ex: "preferences.theme")
        if "." in key:
            parts = key.split(".")
            current = memory
            for i, part in enumerate(parts):
                if i == len(parts) - 1:
                    current[part] = value
                else:
                    if part not in current or not isinstance(current[part], dict):
                        current[part] = {}
                    current = current[part]
        else:
            memory[key] = value
            
        self._save_user_memory(memory)
        return True

    def extract_memories_from_conversation(self, conversation_id: str, messages: List[Dict[str, str]]) -> bool:
        """
        Extrait des informations importantes d'une conversation
        
        Args:
            conversation_id: ID de la conversation
            messages: Liste des messages de la conversation
            
        Returns:
            bool: Succès de l'opération
        """
        memory = self._load_user_memory()
        
        # Extraction des informations personnelles
        for msg in messages:
            if msg["role"] == "user":
                # Extraction du nom
                name_patterns = [
                    r"Je m'appelle ([A-Z][a-z]+)",
                    r"Mon nom est ([A-Z][a-z]+)",
                    r"([A-Z][a-z]+) est mon prénom"
                ]
                
                for pattern in name_patterns:
                    match = re.search(pattern, msg["content"])
                    if match and match.group(1):
                        memory["name"] = match.group(1)
                
                # Extraction des préférences
                preference_patterns = [
                    r"Je préfère ([^.,]+)",
                    r"J'aime ([^.,]+)",
                    r"Ma préférence est ([^.,]+)"
                ]
                
                for pattern in preference_patterns:
                    matches = re.finditer(pattern, msg["content"])
                    for match in matches:
                        if match.group(1):
                            preference = match.group(1).strip().lower()
                            memory["preferences"][f"preference_{len(memory['preferences']) + 1}"] = preference
                
                # Extraction des faits importants
                if "je suis" in msg["content"].lower():
                    facts = re.findall(r"je suis ([^.,]+)", msg["content"].lower())
                    for fact in facts:
                        fact = fact.strip()
                        if fact:
                            memory["facts"][f"fact_{len(memory['facts']) + 1}"] = fact
        
        # Ajout du contexte de la conversation
        conversation_summary = f"Conversation {conversation_id} du {datetime.datetime.now().strftime('%d/%m/%Y')}"
        if conversation_summary not in memory["context"]:
            memory["context"].append(conversation_summary)
            
        self._save_user_memory(memory)
        return True

    def get_memory_context(self) -> str:
        """
        Génère un contexte formaté pour le modèle basé sur les mémoires
        
        Returns:
            str: Contexte formaté
        """
        memory = self._load_user_memory()
        context = "Informations importantes sur l'utilisateur:\n"
        
        if memory["name"]:
            context += f"- Nom: {memory['name']}\n"
        
        if memory["preferences"]:
            context += "- Préférences:\n"
            for key, value in memory["preferences"].items():
                context += f"  * {value}\n"
        
        if memory["facts"]:
            context += "- Faits importants:\n"
            for key, value in memory["facts"].items():
                context += f"  * {value}\n"
        
        if memory["context"]:
            context += "- Contexte supplémentaire:\n"
            for item in memory["context"]:
                context += f"  * {item}\n"
                
        return context

    def get_all_conversations(self) -> Dict[str, Any]:
        """Récupère toutes les conversations enregistrées"""
        conversations = {}
        
        try:
            # Parcourir tous les fichiers dans le répertoire de mémoire
            for filename in os.listdir(MEMORY_DIR):
                # Ne pas inclure le fichier de mémoire utilisateur
                if filename != "user_memory.json" and filename.endswith(".json"):
                    # Extraire l'ID de conversation du nom de fichier
                    conversation_id = filename.replace(".json", "")
                    
                    # Charger la mémoire de la conversation
                    try:
                        memory = self.load_memory(conversation_id)
                        conversations[conversation_id] = memory
                    except Exception as e:
                        print(f"Erreur lors du chargement de la conversation {conversation_id}: {str(e)}")
            
            return conversations
        except Exception as e:
            print(f"Erreur lors de la récupération des conversations: {str(e)}")
            return {}
